pulsar-metadata/src/main/java/org/apache/pulsar/metadata/bookkeeper/PulsarRegistrationClient.java

```diff
@@ -91,38 +96,58 @@ public void close() {
 
     @Override
     public CompletableFuture<Versioned<Set<BookieId>>> getAllBookies() {
         // this method is meant to return all the known bookies, even the bookies
         // that are not in a running state
-        return getChildren(bookieAllRegistrationPath);
+        return getBookiesThenFreshCache(bookieAllRegistrationPath);
     }
 
     @Override
     public CompletableFuture<Versioned<Set<BookieId>>> getReadOnlyBookies() {
-        return getChildren(bookieReadonlyRegistrationPath);
+        return getBookiesThenFreshCache(bookieReadonlyRegistrationPath);
     }
 
-    private CompletableFuture<Versioned<Set<BookieId>>> getChildren(String path) {
+    /**
+     * @throws IllegalArgumentException if parameter path is null or empty.
+     */
+    private CompletableFuture<Versioned<Set<BookieId>>> getBookiesThenFreshCache(String path) {
+        if (path == null || path.isEmpty()) {
+            return failedFuture(
+                    new IllegalArgumentException(String.format("parameter [path] can not be null or empty.")));
+        }
         return store.getChildren(path)
                 .thenComposeAsync(children -> {
-                    Set<BookieId> bookieIds = PulsarRegistrationClient.convertToBookieAddresses(children);
-                    List<CompletableFuture<?>> bookieInfoUpdated =
-                            new ArrayList<>(bookieIds.size());
+                    final Set<BookieId> bookieIds = PulsarRegistrationClient.convertToBookieAddresses(children);
+                    final List<CompletableFuture<?>> bookieInfoUpdated = new ArrayList<>(bookieIds.size());
                     for (BookieId id : bookieIds) {
                         // update the cache for new bookies
-                        if (!bookieServiceInfoCache.containsKey(id)) {
+                        if (path.equals(bookieRegistrationPath) && writableBookieInfo.get(id) == null) {
                             bookieInfoUpdated.add(readBookieServiceInfoAsync(id));
+                            continue;
+                        }
+                        if (path.equals(bookieReadonlyRegistrationPath) && readOnlyBookieInfo.get(id) == null) {
+                            bookieInfoUpdated.add(readBookieInfoAsReadonlyBookie(id));
+                            continue;
+                        }
+                        if (path.equals(bookieAllRegistrationPath)) {
+                            if (writableBookieInfo.get(id) != null || readOnlyBookieInfo != null) {
+                                // jump to next bookie id
+                                continue;
+                            }
+                            bookieInfoUpdated.add(readBookieServiceInfoAsync(id) // check writable first
+                                    .thenCompose(updated ->
+                                            updated ? completedFuture(null)
+                                                    : readBookieInfoAsReadonlyBookie(id))); // check read-only then
```

Comment:
It will be easier to read the code by changing the return type to `Optional<BookieServiceInfo>` for methods `readBookieServiceInfoAsync` and `readBookieInfoAsReadonlyBookie`. So that we can check if the BookieServiceInfo is present or not here. It's a little confusing to use a BOOL value `updated` here.

Reply:
fixed